libs/sdk-core/src/lib/events.ts

```diff
@@ -1,18 +1,29 @@
 import type { Observable } from 'rxjs';
-import { Subject, filter, map } from 'rxjs';
+import { BehaviorSubject, Subject, filter, map } from 'rxjs';
 import type { IEvents } from './models';
 
 export class Events implements IEvents {
+  // instant messages
   private messages = new Subject<{ name: string; data: any }>();
+  // persistent logs
+  private logs = new BehaviorSubject<{ [name: string]: any }>({});
 
-  emit<T>(eventName: string, data: T): void {
+  emit<T = any>(eventName: string, data: T): void {
     this.messages.next({ name: eventName, data });
   }
 
-  on<T>(eventName: string): Observable<T> {
+  on<T = any>(eventName: string): Observable<T> {
     return this.messages.asObservable().pipe(
       filter((message) => message.name === eventName),
       map((message) => message.data as T),
     );
   }
+
+  log(eventName: string, data: any): void {
+    this.logs.next({ ...this.logs.getValue(), [eventName]: data });
```

Comment:
when does the logs reset? won't they grow forever with this code?

Reply:
it's not a list, just a dict, so for any given key, we override the previous value on each change

Reply:
oh right